File: static/wasm/src/bus.cpp

#include "bus.h"
#include "component.h"

uint8_t Bus::Read(uint16_t address) {
    for(int i = 0; i < this->_connections.size(); i++) {
        Connection c = this->_connections[i];
        if(address >= c.addrMin && address <= c.addrMax) {
            return c.component->Read(address, false);
        }
    }
    return 0x00;
}

void Bus::Write(uint16_t address, uint8_t data) {
    for(int i = 0; i < this->_connections.size(); i++) {
        Connection c = this->_connections[i];
        if(address >= c.addrMin && address <= c.addrMax) {
            c.component->Write(address, data);
        }
    }
}

bool Bus::Probe(uint16_t address) {
    for(int i = 0; i < this->_connections.size(); i++) {
        Connection c = this->_connections[i];
        if(address >= c.addrMin && address <= c.addrMax) {
            return true;
        }
    }
    return false;
}

void Bus::Connect(Component * component, uint16_t addr_min, uint16_t addr_max) {
    Connection c;
    c.component = component;
    c.addrMin = addr_min;
    c.addrMax = addr_max;
    this->_connections.push_back(c);
}

File: static/src/data/bus.ts
import { Bitset } from "./bitset";
import { RAM } from "../component/ram";
import { Component } from "../component/component";

interface IConnection {
    component: Component;
    addrMin: number;
    addrMax: number;
}

export class Bus {

    private _connections: IConnection[];

    constructor() {
        this._connections = [];
    }

    public Write(address: number, value: number, rdOnly: boolean = false): void {
        for(let connection of this._connections ) {
            if(address >= connection.addrMin && address <= connection.addrMax) {                
                connection.component.Write(address, value, rdOnly);
                return;
            }
        }
    }

    public Read(address: number, rdOnly: boolean = false): number {
        for(let connection of this._connections ) {
            if(address >= connection.addrMin && address <= connection.addrMax) {                
                return connection.component.Read(address, true);                
            }
        }
        return -1;
    }

    public Connect(component: Component, addrMin: number, addrMax: number, mirrorLength?: number) {
        this._connections.push({
            component: component,
            addrMin: addrMin,
            addrMax: addrMax
        });
    }

}